cranelift/codegen/src/isa/s390x/inst.isle

```diff
@@ -3866,10 +3828,10 @@
 ;; conditional move, and because flogr returns a register pair.
 (rule -1 (clz_reg zeroval x)
       (let ((dst WritableRegPair (temp_writable_regpair))
-            (_ Unit (emit (MInst.Flogr x)))
-            (_ Unit (emit (MInst.CMov64SImm16 (writable_regpair_hi dst)
-                              (intcc_as_cond (IntCC.Equal)) zeroval))))
-        dst))
+            (_ Unit (emit (MInst.Flogr dst x)))
+            (hi WritableReg (temp_writable_reg $I64))
+            (_ Unit (emit (MInst.CMov64SImm16 hi (intcc_as_cond (IntCC.Equal)) (regpair_hi dst) zeroval))))
+        (regpair hi (regpair_lo dst))))
```

Comment:
This also now seems to generate an extra `lgr` (see the `ctz_64` test case).

Reply:
I dug into this a bit, and have an explanation. Here's the program before register allocation:

```
  Inst 0: args %v128=%r2                                                              
  Inst 1: lcgr %v131, %v128                                                           
  Inst 2: ngrk %v132, %v128, %v131                                                    
  Inst 3: flogr %v133/%v134, %v132                                                    
  Inst 4: locghie %v135<-%v133, -1                                                    
  Inst 5: lghi %v136, 63                                                              
  Inst 6: sgrk %v137, %v136, %v135                                                    
  Inst 7: lgr %v130, %v129                                                            
  Inst 8: lgr %r2, %v130                                                              
  Inst 9: br %r14
```

There are two live-ranges that contribute to the move: one that covers the output of instruction 3 to where it's consumed in instruction 4 (`a`), and one that covers the output of instruction 4  to where it's consumed in instruction 6 (`b`). As the second live-range `b` is longer than `a`, it will be allocated first. As it has no fixed register constraints, the allocator is free to choose anything, and in the test output it chooses `%r4`. As the `flogr` instruction has a constraint that says it must use `%r2`/`%r3` for its output register, allocating a register for `%v133` produces a move from `%r2` to `%r4` once the live-range `a` is processed.

It would be great to fix this problem by merging those two live-ranges into a single bundle, but regalloc2 is currently unable to merge live ranges when one contains a fixed register constraint, like `a`. It's not all bad news though, getting cranelift producing lowered output that's in SSA will enable some simplifications in regalloc2 that will enable merging these two live-ranges, so these sort of unnecessary moves should be less common in the future.

Reply:
I see, thanks for having a look.  This looks all good to me now.